provider/operator/ipoperator/ip_operator.go

```diff
@@ -0,0 +1,788 @@
+	kubeConfig, err := clientcommon.OpenKubeConfig(configPath, logger)
+	if err != nil {
+		return err
+	}
+
+	client, err := clusterClient.NewClient(logger, ns, configPath)
+	if err != nil {
+		return err
+	}
+
+	metalLbEndpoint, err := provider_flags.GetServiceEndpointFlagValue(logger, serviceMetalLb)
+	if err != nil {
+		return err
+	}
+
+	mllbc, err := metallb.NewClient(configPath, logger, metalLbEndpoint)
+	if err != nil {
+		return err
+	}
+
+	providerEndpoint, err := provider_flags.GetServiceEndpointFlagValue(logger, serviceProvider)
+	if err != nil {
+		return err
+	}
+
+	providerSda, err := clusterutil.NewServiceDiscoveryAgent(logger, kubeConfig, "gateway", "akash-provider", "akash-services", providerEndpoint)
+	if err != nil {
+		return err
+	}
+	logger.Info("clients", "kube", client, "metallb", mllbc)
+	logger.Info("HTTP listening", "address", listenAddr)
+
+	op, err := newIPOperator(logger, client, operatorcommon.IgnoreListConfigFromViper(), mllbc, providerSda)
+	if err != nil {
+		return err
+	}
+	router := op.server.GetRouter()
+	group, ctx := errgroup.WithContext(cmd.Context())
+
+	group.Go(func() error {
+		srv := http.Server{Addr: listenAddr, Handler: router}
+		go func() {
+			<-ctx.Done()
+			_ = srv.Close()
+		}()
+		err := srv.ListenAndServe()
+		if errors.Is(err, http.ErrServerClosed) {
+			return nil
+		}
+		return err
+	})
+
+	group.Go(func() error {
+		return op.run(ctx)
+	})
+
+	err = group.Wait()
+	if err != nil && !errors.Is(err, context.Canceled) && !errors.Is(err, http.ErrServerClosed) {
+		return err
+	}
+	return nil
+}
+
+func Cmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:          "ip-operator",
+		Short:        "kubernetes operator interfacing with Metal LB",
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return doIPOperator(cmd)
+		},
+	}
+	operatorcommon.AddOperatorFlags(cmd, "0.0.0.0:8086")
+	operatorcommon.AddIgnoreListFlags(cmd)
+
+	if err := provider_flags.AddServiceEndpointFlag(cmd, serviceProvider); err != nil {
+		return nil
+	}
+
+	if err := provider_flags.AddServiceEndpointFlag(cmd, serviceMetalLb); err != nil {
+		return nil
+	}
+
+	return cmd
+}
+
+func (op *ipOperator) run(parentCtx context.Context) error {
+	op.log.Debug("ip operator start")
+	const threshold = 500 * time.Millisecond
+	for {
+		lastAttempt := time.Now()
+		err := op.monitorUntilError(parentCtx)
+		if errors.Is(err, context.Canceled) {
+			op.log.Debug("ip operator terminate")
+			break
+		}
+
+		op.log.Error("observation stopped", "err", err)
+
+		// don't spin if there is a condition causing fast failure
+		elapsed := time.Since(lastAttempt)
+		if elapsed < threshold {
+			op.log.Info("delaying")
+			select {
+			case <-parentCtx.Done():
+				break
+
+			case <-time.After(threshold):
+				// delay complete
+			}
+		}
+	}
+
+	op.providerSda.Stop()
+	op.mllbc.Stop()
+	return parentCtx.Err()
+}
+
+type barrier struct {
```

Comment:
I think this type has a race in it, but I am not sure. Need to double check.

Reply:
It looks like storing a value, when I meant to subtract that value. This has been corrected.